parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/ProducerManager.java

```diff
@@ -114,42 +114,39 @@ private TransactionManager getTransactionManager() {
     }
 
     /**
-     * Produce a message back to the broker.
-     * <p>
-     * Implementation uses the blocking API, performance upgrade in later versions, is not an issue for the more common
-     * use case where messages aren't produced.
+     * Produce messages back to the broker.
+     *
+     * <p>Produced messages are not waited for, instead their futures are returned so the caller can
+     * decide when waiting should be done (if at all).
      *
      * @see ParallelConsumer#poll
      * @see ParallelStreamProcessor#pollAndProduceMany
      */
-    public RecordMetadata produceMessage(ProducerRecord<K, V> outMsg) {
+    public List<ParallelConsumer.Tuple<ProducerRecord<K, V>, Future<RecordMetadata>>> produceMessages(
+            List<ProducerRecord<K, V>> outMsgs) {
         // only needed if not using tx
-        Callback callback = (RecordMetadata metadata, Exception exception) -> {
-            if (exception != null) {
-                log.error("Error producing result message", exception);
-                throw new RuntimeException("Error producing result message", exception);
-            }
-        };
+        Callback callback =
+                (RecordMetadata metadata, Exception exception) -> {
+                    if (exception != null) {
+                        log.error("Error producing result message", exception);
+                        throw new RuntimeException("Error producing result message", exception);
+                    }
+                };
 
         ReentrantReadWriteLock.ReadLock readLock = producerTransactionLock.readLock();
         readLock.lock();
-        Future<RecordMetadata> send;
+        List<ParallelConsumer.Tuple<ProducerRecord<K, V>, Future<RecordMetadata>>> futures =
+                new ArrayList<>(outMsgs.size());
         try {
-            send = producer.send(outMsg, callback);
+            for (ProducerRecord<K, V> record : outMsgs) {
```

Comment:
```suggestion
            for (ProducerRecord<K, V> rec: outMsgs) {
```
or similar - `record` is a restricted identifier

Reply:
(check out the solar lint plugin for IDEA)

Reply:
=== Recommended IDEA Plugins

* AsciiDoc
* CheckStyle
* CodeGlance
* Rainbow Brackets
* SonarLint
* Lombok

Reply:
fixed